app/models/labware_creators/common_file_handling/csv_file_for_tube_rack.rb

```diff
@@ -0,0 +1,127 @@
+    NO_DUPLICATE_TUBE_BARCODES_MSG = 'contains duplicate tube barcodes (%s)'
+
+    #
+    # Extracts tube details by rack location from the uploaded csv file.
+    # This hash is useful when we want the details for a rack location.
+    #
+    # @return [Hash] e.g. { 'A1' => { details for this location }, 'B1' => etc. }
+    #
+    def position_details
+      @position_details ||= generate_position_details_hash
+    end
+
+    # This hash is useful when we want to know the location of a tube barcode
+    # within the rack.
+    #
+    # @return [Hash] eg. { 'FX00000001' => 'A1', 'FX00000002 => 'B1' etc. }
+    #
+    def location_by_barcode_details
+      @location_by_barcode_details ||=
+        position_details.each_with_object({}) { |(position, details), hash| hash[details['tube_barcode']] = position }
+    end
+
+    private
+
+    # Returns an array of Row objects representing the tube rack scan data in the CSV file.
+    #
+    # @return [Array<Row>] An array of Row objects.
+    def tube_rack_scan
+      @tube_rack_scan ||=
+        @data[0..].each_with_index.map do |row_data, index|
+          CommonFileHandling::CsvFile::RowForTubeRack.new(index, row_data)
+        end
+    end
+
+    # Checks for duplicate rack positions in the tube rack scan.
+    # If any duplicates are found, an error message is added to the errors object.
+    # The error message includes the duplicated rack posotions.
+    # This method is used to ensure that each rack position in the tube rack scan is unique.
+    def check_no_duplicate_rack_positions
+      duplicated_rack_positions =
+        tube_rack_scan.group_by(&:tube_position).select { |_position, tubes| tubes.size > 1 }.keys.join(',')
+
+      return if duplicated_rack_positions.empty?
+
+      errors.add(:base, format(NO_DUPLICATE_RACK_POSITIONS_MSG, duplicated_rack_positions))
+    end
+
+    # Checks for duplicate tube barcodes in the tube rack scan.
+    # If any duplicates are found, they are added to the errors object.
+    # The error message includes the duplicated tube barcodes.
+    # 'NO READ' and 'NOSCAN' values are ignored and not considered as duplicates.
+    # This method is used to ensure that each tube barcode in the tube rack scan is unique.
+    def check_no_duplicate_tube_barcodes
+      duplicates = tube_rack_scan.group_by(&:tube_barcode).select { |_tube_barcode, tubes| tubes.size > 1 }.keys
+
+      # remove any NO READ or NOSCAN or empty string values from the duplicates
+      duplicates = duplicates.reject { |barcode| NO_TUBE_TEXTS.include?(barcode) }
+
+      return if duplicates.empty?
+
+      errors.add(:base, format(NO_DUPLICATE_TUBE_BARCODES_MSG, duplicates.join(',')))
+    end
+
+    # Generates a hash of position details based on the tube rack scan data in the CSV file.
+    #
+    # @return [Hash] A hash of position details, where the keys are positions and the values
+    # are hashes containing the tube rack barcode and tube barcode for each position.
```

Comment:
```suggestion
    # are hashes containing the tube barcode for each position.
```

Is this correct? Does not contain the tube rack barcode?

Reply:
correct, fixed